File: agents/javascript/src/api-client.ts
/**
 * HTTP client for the Mock API server.
 * Uses native Node.js http module — no external dependencies.
 */

import * as http from 'http';
import * as https from 'https';
import { Message } from './state';

export interface ChatCompletionPayload {
  model: string;
  messages: Message[];
  tools: unknown[];
}

export async function sendChatCompletion(
  apiUrl: string,
  sessionId: string,
  payload: ChatCompletionPayload
): Promise<string> {
  const url = new URL('/v1/chat/completions', apiUrl);
  const body = JSON.stringify(payload);
  const isHttps = url.protocol === 'https:';

  return new Promise((resolve, reject) => {
    const options: http.RequestOptions = {
      hostname: url.hostname,
      port: url.port,
      path: url.pathname,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
        'Authorization': `Bearer ${sessionId}`,
        'X-Session-Id': sessionId,
      },
    };

    const req = (isHttps ? https : http).request(options, (res) => {
      const chunks: Buffer[] = [];
      res.on('data', (chunk: Buffer) => chunks.push(chunk));
      res.on('end', () => {
        const responseBody = Buffer.concat(chunks).toString('utf-8');
        if (res.statusCode && res.statusCode >= 400) {
          reject(new Error(`API returned ${res.statusCode}: ${responseBody}`));
        } else {
          resolve(responseBody);
        }
      });
    });

    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

File: agents/rust/src/api_client.rs
use crate::types::{ChatCompletionRequest, Message};

pub struct ApiClient {
    base_url: String,
    session_id: String,
}

impl ApiClient {
    pub fn new(base_url: String, session_id: String) -> Self {
        Self { base_url, session_id }
    }

    pub fn send_chat_completion(
        &self,
        messages: &[Message],
        tools: &serde_json::Value,
    ) -> Result<String, String> {
        let request = ChatCompletionRequest {
            model: "gpt-4".to_string(),
            messages: messages.to_vec(),
            tools: tools.clone(),
        };

        let body = serde_json::to_string(&request)
            .map_err(|e| format!("serialize request: {}", e))?;

        let url = format!("{}/v1/chat/completions", self.base_url);

        let resp = ureq::post(&url)
            .set("Content-Type", "application/json")
            .set("Authorization", &format!("Bearer {}", self.session_id))
            .set("X-Session-Id", &self.session_id)
            .send_string(&body)
            .map_err(|e| format!("send request: {}", e))?;

        resp.into_string()
            .map_err(|e| format!("read response: {}", e))
    }
}
